### PR TITLE
Start of pycbc qtransform inspiral integration

### PR DESCRIPTION
**Pull request by hagabbar**
pycbc_inspiral changes:
1.) added import of qtransform.py script from pycbc.filter
2.) generate qtransform for each segment in a pycbc_inspiral run
- Still need to save output to ref.hdf file
-Make output compatible with pycbc_coinc_mergetrigs module

qtransform.py changes:
1.) qplane function now returns uninterpolated qtransforms by default
2.) functions for use of qtransform.py in pycbc_inspiral have been added
3.) new functions to recursively save/convert multilevel python dictionary to h5py file and vice versa.

### CODE REVIEW

File: pycbc/filter/qtransform.py [hunk 1]
```diff
@@ -42,7 +42,160 @@
...
+    Returns
+    -------
+    Qbase: 'dict'
+        dictionary containing Q-tile tuples for a set of Q-planes
+    frange: 'list'
+        upper and lower bounds on frequency range
+    data:
+        whitened strain of analysis chunk
+
+    """
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Why are you recalculating a PSD here, instead of just using the ones we've already calculated?

File: pycbc/filter/qtransform.py [hunk 2]
```diff
@@ -42,7 +42,160 @@
 from pycbc.types import zeros
 from numpy import fft as npfft
 import os
+import h5py
+from pycbc.filter import highpass_fir, matched_filter
+from pycbc.psd import welch, interpolate
 
+def pycbc_insp_main(segments, filename='q_info.hdf'):
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Please give things a bit more meaningful names, not tied to a specific executable. Presumably this function could be used elsewhere.

File: pycbc/filter/qtransform.py (hunk 1 above)
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: On further look, it looks like your code is broken, as you pass the psd to this function, but it doesn't accept a psd.... can you check that you have committed the correct things?

File: pycbc/filter/qtransform.py [hunk 3]
```diff
@@ -42,7 +42,160 @@
...
+    # retrieve whitened strain
+    white_strain = (data.to_frequencyseries() / seg_psd ** 0.5 * seg_psd.delta_f)            
+    data = white_strain
+ 
+    # perform Q-tiling
+    Qbase, frange = qtiling(data, qrange, frange, mismatch)
+
+    return Qbase, frange, data
+
+def save_dict_to_hdf5(dic, filename):
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: This and the following functions are pretty generic. If we need them, please place in pycbc/io/hdf.py

File: pycbc/filter/qtransform.py (hunk 1 above)
Comment (hagabbar) [inline pycbc/filter/qtransform.py]: Well if I use the psd from the original segments the precisions won't match when computing the whitened strain.

Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Well, do you need to do either of those things?

Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Why are you doing another highpass here for example?

Comment (hagabbar) [inline pycbc/filter/qtransform.py]: @ahnitz I went ahead and removed the additional highpass filter, removed the recalculation of the psd line, changed a few variable names in the function.

File: pycbc/filter/qtransform.py (hunk 3 above)
Comment (hagabbar) [inline pycbc/filter/qtransform.py]: Moved.

File: pycbc/filter/qtransform.py (hunk 2 above)
Comment (hagabbar) [inline pycbc/filter/qtransform.py]: renamed.

File: pycbc/filter/qtransform.py [hunk 4]
```diff
@@ -42,7 +42,88 @@
...
+    data:
+        whitened strain of analysis chunk
+
+    """
+
+    # assume segment is fft'd
+    # check for sampling rate
+    sampling = (len(seg) - 1) * 2 * seg.delta_f
+
+    data = seg.to_timeseries()
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: So, why are you converting this to the time domain, only to convert it bank?

File: pycbc/filter/qtransform.py [hunk 5]
```diff
@@ -42,7 +42,88 @@
...
+
+    # highpass and suppress frequencies below 20Hz
+    dur = data.duration
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: This is unused.

File: pycbc/filter/qtransform.py [hunk 6]
```diff
@@ -128,7 +209,7 @@ def qplane(qplane_tile_dict, fseries, frange, normalized=True, tres=0.001, fres=
     """
     # store q-transforms of each tile in a dict
     qplane_qtrans_dict = {}
-    dur = 1.0 / fseries.delta_f
+    dur = fseries.to_timeseries().duration
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: frequency series have a duration property now, please use that instead of converting to the time domain.

File: pycbc/filter/qtransform.py [hunk 7]
```diff
@@ -128,7 +209,7 @@ def qplane(qplane_tile_dict, fseries, frange, normalized=True, tres=0.001, fres=
...
 
     # check for sampling rate
     sampling = (len(fseries) - 1) * 2 * fseries.delta_f
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: fseris also has a sample_rate property now.

File: pycbc/filter/qtransform.py [hunk 8]
```diff
@@ -200,7 +300,7 @@ def qtiling(fseries, qrange, frange, mismatch):
     deltam = deltam_f(mismatch)
     qrange = (float(qrange[0]), float(qrange[1]))
     frange = [float(frange[0]), float(frange[1])]
-    dur = 1.0 / fseries.delta_f
+    dur = fseries.to_timeseries().duration
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: same here, don't convert to time domain. In fact, it was OK before, why did you change it?

File: pycbc/filter/qtransform.py [hunk 9]
```diff
@@ -315,18 +415,16 @@ def qtransform(fseries, Q, f0):
         A 'TimeSeries' of the complex energy from the Q-transform of 
         this tile against the data.
     """
+
     # q-transform data for each (Q, frequency) tile
+
     # initialize parameters
     qprime = Q / 11**(1/2.) # ... self.qprime
-    dur = 1.0 / fseries.delta_f
+    dur = fseries.to_timeseries().duration
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: and here, same issue as earlier.

File: pycbc/filter/qtransform.py [hunk 10]
```diff
@@ -337,20 +435,29 @@ def qtransform(fseries, Q, f0):
...
+    output_samples = dur * output_sampling
 
     # pad data, move negative frequencies to the end, and IFFT
     padded = np.pad(windowed, padding(window_size, output_samples), mode='constant')
     wenergy = npfft.ifftshift(padded)
 
     # return a 'TimeSeries'
     wenergy = FrequencySeries(wenergy, delta_f=1./dur)
-    cenergy = TimeSeries(zeros(output_samples, dtype=np.complex128),
+    tdenergy = TimeSeries(zeros(int(output_samples), dtype=np.complex128),
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Why did you make these changes? The existing version seems better... and faster.

File: pycbc/io/hdf.py [hunk 11]
```diff
@@ -835,3 +835,62 @@ def get_chisq_from_file_choice(hdfile, chisq_choice):
         raise ValueError(err_msg)
     return chisq
 
+def save_dict_to_hdf5(dic, filename):
```
Comment (ahnitz) [inline pycbc/io/hdf.py]: Where is this called from? Is that not implemented yet? Just checking you are missing commits.

File: bin/pycbc_inspiral [hunk 12]
```diff
@@ -307,13 +308,16 @@ with ctx:
 
     tsetup = time.time() - tstart
 
+    logging.info("Performing q-transform on analysis segments")
+    out_vals['qtransform'] = qtransform.inspiral_qtransform_generator(segments)
```
Comment (ahnitz) [inline bin/pycbc_inspiral]: Does this actually work? If so, interesting, that's not intended but not terrible I guess.

Comment (ahnitz) [inline bin/pycbc_inspiral]: I.E. does it actually save out the results correctly? Also, we don't always want to have these results. please add a command line arguments to toggle the qtransform related calls. If this code actually functions, you can update the dictionary above and put both these lines into an if statement.

File: bin/pycbc_inspiral [hunk 13]
```diff
@@ -223,6 +223,7 @@ with ctx:
         'bank_chisq'     : float32,
         'bank_chisq_dof' : int,
         'cont_chisq'     : float32,
+        'qtransform'     : dict,
```
Comment (ahnitz) [inline bin/pycbc_inspiral]: Does this actually work?

**Comment by ahnitz** [comment]
@hagabbar Lots more comments, also please check the code quality checks after you've addressed my comments.

File: bin/pycbc_inspiral (hunk 12 above)
Comment (hagabbar) [inline bin/pycbc_inspiral]: Nope doesn't actually save anything in the output. That was just my latest attempt at trying to figure out how to store the qtransform results in ref.hdf.

File: pycbc/io/hdf.py (hunk 11 above)
Comment (hagabbar) [inline pycbc/io/hdf.py]: As of right now it's not called from anything. But I thought it might be useful in the future for saving the output of the qtransform results.

File: pycbc/filter/qtransform.py (hunk 10 above)
Comment (hagabbar) [inline pycbc/filter/qtransform.py]: Unfortunately, I don't quite remember what the issue was before. But I could do a test run with the old version and see if anything fails.

File: bin/pycbc_inspiral (hunk 13 above)
Comment (hagabbar) [inline bin/pycbc_inspiral]: As of right now, no. But it also doesn't affect the output of the code in any detrimental way.

File: pycbc/filter/qtransform.py [hunk 14]
```diff
@@ -43,6 +43,80 @@
...
+        dictionary containing Q-tile tuples for a set of Q-planes
+    frange: 'list'
+        upper and lower bounds on frequency range
+    data:
+        whitened strain of analysis chunk
+
+    """
+
+    # assume segment is fft'd
+    data = seg.to_timeseries()
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: This is not needed, all you do is immediately invert the transform.

File: pycbc/filter/qtransform.py [hunk 15]
```diff
@@ -43,6 +43,80 @@
...
+
+    # retrieve segment psd value
+    seg_psd = seg.psd
+
+    # retrieve whitened strain
+    white_strain = (data.to_frequencyseries() / seg_psd ** 0.5 * seg_psd.delta_f)
+    data = white_strain
+ 
+    # perform Q-tiling
+    Qbase, frange = qtiling(data, qrange, frange, mismatch)
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Try to avoid capitals in names.

File: pycbc/filter/qtransform.py [hunk 16]
```diff
@@ -337,20 +427,29 @@ def qtransform(fseries, Q, f0):
...
     # return a 'TimeSeries'
     wenergy = FrequencySeries(wenergy, delta_f=1./dur)
-    cenergy = TimeSeries(zeros(output_samples, dtype=np.complex128),
+    tdenergy = TimeSeries(zeros(output_samples, dtype=np.complex128),
                             delta_t=1./sampling)
-    ifft(wenergy, cenergy)
-    energy = cenergy.squared_norm()
-    medianenergy = np.median(energy.numpy())
-    norm_energy = energy / float(medianenergy)
+    ifft(wenergy, tdenergy)
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Why are you making these changes?

Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Specifically lines 442 - 451.

File: pycbc/filter/qtransform.py [hunk 17]
```diff
@@ -43,6 +43,80 @@
...
+
+        # getting q-plane for segment
+        Qplane, interp, qs_time, qe_time = qplane(Qbase, q_data, q_frange, fres=None, seg=stilde)
+        
+        del q_frange, q_data, interp
+
+        # write q info to an hdf file
+        Qbase_tmp = {}
+        Qplane_tmp = {}
+        print 'start time: %s, end time: %s' % (qs_time, qe_time)
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Try not to use a print statement. If you want logging, use the logging module.

File: pycbc/filter/qtransform.py [hunk 18]
```diff
@@ -43,6 +43,80 @@
...
+        del s_num
+        Qbase, q_frange, q_data = inspiral_tiling(stilde)
+
+        # getting q-plane for segment
+        Qplane, interp, qs_time, qe_time = qplane(Qbase, q_data, q_frange, fres=None, seg=stilde)
+        
+        del q_frange, q_data, interp
+
+        # write q info to an hdf file
+        Qbase_tmp = {}
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: avoid capitals in naming (PEP8)

File: pycbc/filter/qtransform.py (hunk 16 above)
Comment (hagabbar) [inline pycbc/filter/qtransform.py]: You mean why did I change the name of the variable cenergy?

Comment (ahnitz) [inline pycbc/filter/qtransform.py]: You also added in an extra line, and removed the usage of a method for a slower and more complicated function.

Comment (hagabbar) [inline pycbc/filter/qtransform.py]: Hmm, yeah I'm not really sure why these changes were made.

Comment (hagabbar) [inline pycbc/filter/qtransform.py]: I'm testing out the old version right now

Comment (hagabbar) [inline pycbc/filter/qtransform.py]: Yep everything seems to work fine, so I'll go ahead and switch to the older, faster version.

File: bin/pycbc_inspiral (hunk 13 above)
Comment (hagabbar) [inline bin/pycbc_inspiral]: removed line 229

File: pycbc/filter/qtransform.py (hunk 18 above)
Comment (hagabbar) [inline pycbc/filter/qtransform.py]: done

File: pycbc/filter/qtransform.py (hunk 17 above)
Comment (hagabbar) [inline pycbc/filter/qtransform.py]: removed

File: bin/pycbc_inspiral [hunk 19]
```diff
@@ -72,6 +72,9 @@ parser.add_argument("--max-template-length", type=float,
                   help="The maximum length of a template is seconds. The "
                        "starting frequency of the template is modified to "
                        "ensure the proper length")
+parser.add_argument("--q-transform", type=bool, default=False,
```
Comment (ahnitz) [inline bin/pycbc_inspiral]: Make option --enable-q-transform. Use action='store_true', then remove the type and default.

File: pycbc/filter/qtransform.py (hunk 15 above)
Comment (hagabbar) [inline pycbc/filter/qtransform.py]: done

File: bin/pycbc_inspiral [hunk 20]
```diff
@@ -238,10 +241,17 @@ with ctx:
         logging.info("Finished")
         sys.exit(0)
 
+    if opt.q_transform == True:
```
Comment (ahnitz) [inline bin/pycbc_inspiral]: You can just write if opt.enable_q_transform: also the next line should be an else. (What other options are there??).

Comment (hagabbar) [inline bin/pycbc_inspiral]: Ooops, yes good catch

File: pycbc/events/events.py [hunk 21]
```diff
@@ -731,6 +742,18 @@ def __setitem__(self, name, data):
                 f['search/setup_time_fraction'] = \
                     numpy.array([float(self.setup_time) / float(self.run_time)])
 
+            if 'q_trans' in self.global_params:
```
Comment (ahnitz) [inline pycbc/events/events.py]: Let's not add this to them multievent case. You aren't adding into anything that will call this.
